### ISSUE TITLE
Compile error for test_loop11 in onnx backend test

### ISSUE THREAD

**Issue by chentong319** [onnx/onnx-mlir#1236]
The error message is 
`'affine.for' op operand cannot be used as a symbol`
With further investigation, the error came from the upper bound (%43) of this loop:
```
    %43 = arith.select %42, %c0, %41 : index
    %44 = memref.alloc(%43) {alignment = 16 : i64} : memref<?xf32>
    affine.for %arg4 = 0 to %43 {
```
The reason is that this loop is inside the affine loop for the LoopOp.  Symbols should satisfy the [restrictions of affine dialect](https://mlir.llvm.org/docs/Dialects/Affine/#restrictions-on-dimensions-and-symbols). My feeling is that we can not naively generate nested affine loop from nested Krnl loop. For the simplified example, the upper bound, though is a loop invariant of the outer loop, is computed inside the outer loop. 
I do not have a general solution yet. A fix for LoopOp: add an region with trait AffineScope to enclose the body of LoopOp. This fix is just to avoid the error for the loop nested introduced by the lowering of LoopOp, but can not guarantee correctly lowering arbitrary krnl loop structure. 
A side question, will IndexExpr with the right insertion point help in generating the code at the right place?
Here is the EmitONNXIR output of test_loop11_cpu:
```
module attributes {llvm.data_layout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", llvm.target_triple = "x86_64-apple-darwin20.3.0"} {
  func @main_graph(%arg0: tensor<i64>, %arg1: tensor<i1>, %arg2: tensor<1xf32>) -> (tensor<1xf32>, tensor<?x1xf32>) attributes {input_names = ["trip_count", "cond", "y"], output_names = ["res_y", "res_scan"]} {
    %0:2 = "onnx.Loop"(%arg0, %arg1, %arg2) ({
    ^bb0(%arg3: tensor<i64>, %arg4: tensor<i1>, %arg5: tensor<1xf32>):
      %1 = "onnx.Constant"() {value = dense<[1.000000e+00, 2.000000e+00, 3.000000e+00, 4.000000e+00, 5.000000e+00]> : tensor<5xf32>} : () -> tensor<5xf32>
      %2 = "onnx.Constant"() {value = dense<1> : tensor<i64>} : () -> tensor<i64>
      %3 = "onnx.Add"(%arg3, %2) : (tensor<i64>, tensor<i64>) -> tensor<i64>
      %4 = "onnx.UnsqueezeV11"(%arg3) {axes = [0]} : (tensor<i64>) -> tensor<1xi64>
      %5 = "onnx.UnsqueezeV11"(%3) {axes = [0]} : (tensor<i64>) -> tensor<1xi64>
      %6 = "onnx.Constant"() {value = dense<0> : tensor<1xi64>} : () -> tensor<1xi64>
      %7 = "onnx.Constant"() {value = dense<1> : tensor<1xi64>} : () -> tensor<1xi64>
      %8 = "onnx.Slice"(%1, %4, %5, %6, %7) : (tensor<5xf32>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<?xf32>
      %9 = "onnx.Add"(%arg5, %8) : (tensor<1xf32>, tensor<?xf32>) -> tensor<1xf32>
      onnx.Return %arg4, %9, %9 : tensor<i1>, tensor<1xf32>, tensor<1xf32>
    }) {input_names = ["iter_count", "cond_in", "y_in"], output_names = ["cond_out", "y_out", "scan_out"]} : (tensor<i64>, tensor<i1>, tensor<1xf32>) -> (tensor<1xf32>, tensor<?x1xf32>)
    return %0#0, %0#1 : tensor<1xf32>, tensor<?x1xf32>
  }
  "onnx.EntryPoint"() {func = @main_graph, numInputs = 3 : i32, numOutputs = 2 : i32, signature = "[    { \22type\22 : \22i64\22 , \22dims\22 : [] , \22name\22 : \22trip_count\22 }\0A ,    { \22type\22 : \22i1\22 , \22dims\22 : [] , \22name\22 : \22cond\22 }\0A ,    { \22type\22 : \22f32\22 , \22dims\22 : [1] , \22name\22 : \22y\22 }\0A\0A]\00@[   { \22type\22 : \22f32\22 , \22dims\22 : [1] , \22name\22 : \22res_y\22 }\0A ,    { \22type\22 : \22f32\22 , \22dims\22 : [5 , 1] , \22name\22 : \22res_scan\22 }\0A\0A]\00"} : () -> ()
}
```

Here is the simplified example:
```
module attributes {llvm.data_layout = "e-m:o-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128", llvm.target_triple = "x86_64-apple-darwin20.3.0"} {
  func @main_graph(%arg0: tensor<i64>, %arg1: tensor<i1>, %arg2: tensor<1xi64>) -> tensor<?x?xf32> attributes {input_names = ["trip_count", "cond", "y"], output_names = [ "res_scan"]} {
    %0 = "onnx.Loop"(%arg0, %arg1) ({
    ^bb0(%arg3: tensor<i64>, %arg4: tensor<i1>):
      %1 = "onnx.Constant"() {value = dense<[1.000000e+00, 2.000000e+00, 3.000000e+00, 4.000000e+00, 5.000000e+00]> : tensor<5xf32>} : () -> tensor<5xf32>
      %6 = "onnx.Constant"() {value = dense<0> : tensor<1xi64>} : () -> tensor<1xi64>
      %7 = "onnx.Constant"() {value = dense<1> : tensor<1xi64>} : () -> tensor<1xi64>
      %9 = "onnx.Slice"(%1, %arg2, %7, %6, %7) : (tensor<5xf32>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>, tensor<1xi64>) -> tensor<?xf32>
      onnx.Return %arg4, %9 : tensor<i1>, tensor<?xf32>
    }) {input_names = ["iter_count", "cond_in", "y_in"], output_names = ["cond_out", "y_out", "scan_out"]} : (tensor<i64>, tensor<i1>) -> tensor<?x?xf32>
    return %0 : tensor<?x?xf32>
  }
  "onnx.EntryPoint"() {func = @main_graph, numInputs = 3 : i32, numOutputs = 1 : i32, signature = "[    { \22type\22 : \22i64\22 , \22dims\22 : [] , \22name\22 : \22trip_count\22 }\0A ,    { \22type\22 : \22i1\22 , \22dims\22 : [] , \22name\22 : \22cond\22 }\0A ,    { \22type\22 : \22f32\22 , \22dims\22 : [1] , \22name\22 : \22y\22 }\0A\0A]\00@[  { \22type\22 : \22f32\22 , \22dims\22 : [-1 , -1] , \22name\22 : \22res_scan\22 }\0A\0A]\00"} : () -> ()
}
```

**Comment by tungld** [comment]
We have been calling `loop-invariant-code-motion` pass at the beginning of `krnl-to-affine` to hoist loop invariant code outside of the loop (https://github.com/onnx/onnx-mlir/blob/main/src/Conversion/KrnlToAffine/ConvertKrnlToAffine.cpp#L601-L607). But it seems it does not work with this example, perhaps because of the code being inside `scf.if`.

> My feeling is that we can not naively generate nested affine loop from nested Krnl loop.

Yes, krnl loop is more generic than affine loop. We would need `scf.for` too.

**Comment by chentong319** [comment]
In the test_loop11, the loop bound is changing with loop induction variable.
In the simplified test, the loop bound is a load of an argument of the function. I agree that scf.if, as well as  the outer loop, may prevent this load and the related computation from being hoisted out.

**Comment by chentong319** [comment]
Updates:
I tried to use scf.for for the LoopOp and got the same error on another loop. This loop nests is to convert the memref<memref> to memref after LoopOp. The complaint is on the upper bound of the inner loop, which is not affine.
```
    krnl.iterate(%11) with (%11 -> %arg3 = 0 to %7){
      %12 = krnl.load %2[%arg3] : memref<?xmemref<?xf32>>
      %13 = krnl.define_loops 1
      %c0_2 = arith.constant 0 : index
      %14 = memref.dim %12, %c0_2 : memref<?xf32>
      krnl.iterate(%13) with (%13 -> %arg4 = 0 to %14){
        %15 = krnl.load %12[%arg4] : memref<?xf32>
        krnl.store %15, %10[%arg3, %arg4] : memref<?x?xf32>
      }
    }
```
The inner loop is generated with utility function emitCopy. It is possible to move the code for loop bound out if the code is generated manually with the knowledge of the whole loop nests and the ONNX semantics: all the tensor in %2 should have the same same. But we will lose the advantages of krnl if we have to generate code in that way.
A quick hack could be: introduce a compiler option or krnl loop flag so that krnl loop could be converted to scf.loop.

**Comment by AlexandreEichenberger** [comment]
@chentong319  Could you give us a copy of the whole code so that we may try to compile and experiment?

**Comment by chentong319** [comment]
> @chentong319 Could you give us a copy of the whole code so that we may try to compile and experiment?

You can find the content of the .mlir in this thread.

**Comment by chentong319** [comment]
Update:

1. We need to add a region with AffineScope if a symbol is not defined at the top of the nearest existing region.
2. The region introduced by SCF ops or Krnl.iterate do not have AffineScope trait. The only operation in MLIR dialect with AffineScope trait is FuncOp.

3. It is obvious that we need to either do outlining to use FuncOp, or define our own op with AffineScope trait to mark the boundary of affine scope. 

I tried with KrnlRegionOp.
I attached the
[slides](https://github.com/onnx/onnx-mlir/files/8375968/03-29-2022-weekly.pptx) 
The simplified test_loop11, called loop11-dyn.mlir
[loop11-dyn.txt](https://github.com/onnx/onnx-mlir/files/8375986/loop11-dyn.txt)
The mlir file after lowering to Krnl with new KrnlRegionOp.
[afterkrnl.txt](https://github.com/onnx/onnx-mlir/files/8375997/afterkrnl.txt)
The mlir file after KrnlRegionOp is lowered, which should be done before convert-scf-to-cf
[beforescftocontrolflow.txt](https://github.com/onnx/onnx-mlir/files/8375999/beforescftocontrolflow.txt)
Note: all .mlir is changed to .txt for git uploading.
